domain/application/state/application.go

```diff
@@ -3014,6 +3016,282 @@ WHERE application_uuid = $applicationID.uuid;
+	insertAppConstraintsStmt, err := st.Prepare(insertAppConstraintsQuery, setApplicationConstraint{})
+	if err != nil {
+		return internalerrors.Errorf("preparing insert application constraints query: %w", err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var containerTypeID containerTypeID
+		if cons.Container != nil {
+			err := tx.Query(ctx, selectContainerTypeIDStmt, containerTypeVal{Value: string(*cons.Container)}).Get(&containerTypeID)
+			if errors.Is(err, sql.ErrNoRows) {
+				return applicationerrors.InvalidApplicationConstraints
+			}
+			if err != nil {
+				return internalerrors.Capture(err)
+			}
+		}
+
+		constraints := encodeConstraints(constraintUUID.String(), cons, containerTypeID.ID)
+
+		if err := tx.Query(ctx, insertConstraintsStmt, constraints).Run(); err != nil {
+			return internalerrors.Capture(err)
+		}
+
+		if cons.Tags != nil {
+			for _, tag := range *cons.Tags {
+				constraintTag := setConstraintTag{ConstraintUUID: constraintUUID.String(), Tag: tag}
+				if err := tx.Query(ctx, insertConstraintTagsStmt, constraintTag).Run(); err != nil {
+					return internalerrors.Capture(err)
+				}
+			}
+		}
+
+		if cons.Spaces != nil {
+			for _, space := range *cons.Spaces {
+				constraintSpace := setConstraintSpace{ConstraintUUID: constraintUUID.String(), Space: space}
+				err := tx.Query(ctx, insertConstraintSpacesStmt, constraintSpace).Run()
+				// Space is not found.
+				if internaldatabase.IsErrConstraintForeignKey(err) {
+					return applicationerrors.InvalidApplicationConstraints
+				}
+				if err != nil {
+					return internalerrors.Capture(err)
+				}
+			}
+		}
+
+		if cons.Zones != nil {
+			for _, zone := range *cons.Zones {
+				constraintZone := setConstraintZone{ConstraintUUID: constraintUUID.String(), Zone: zone}
+				if err := tx.Query(ctx, insertConstraintZonesStmt, constraintZone).Run(); err != nil {
+					return internalerrors.Capture(err)
+				}
+			}
+		}
+
+		return internalerrors.Capture(
+			tx.Query(ctx, insertAppConstraintsStmt, setApplicationConstraint{
+				ApplicationUUID: appID.String(),
+				ConstraintUUID:  constraintUUID.String(),
+			}).Run(),
+		)
+	})
+
+}
+
+// decodeConstraints flattens and maps the list of rows of applicatioConstraint
+// to get a single constraints.Value. The flattening is needed because of the
+// spaces, tags and zones constraints which are slices. We can safely assume
+// that the non-slice values are repeated on every row so we can safely
+// overwrite the previous value on each iteration.
+func decodeConstraints(cons applicationConstraints) constraints.Value {
+	res := constraints.Value{}
+
+	// Empty constraints is not an error case, so return early the empty
+	// result.
+	if len(cons) == 0 {
+		return res
+	}
+
+	// Unique spaces, tags and zones:
+	spaces := make(map[string]string)
+	tags := make(map[string]string)
+	zones := make(map[string]string)
+
+	for _, row := range cons {
```

Comment:
Using `Get` and then passing a singular `applicationConstraint` here will remove the need for this weird for loop. Each interaction here over-writes the previous, which is not obvious, and probably not deterministic since I don't think sql would guarentee the order if multiple `applicationConstraint`s are returned

Reply:
This is related with your comment above, but your reasoning is not correct. Since there are embedded slices in the constraints, their values come from the JOINs with other tables, therefore I have a list of rows that I need to unwrap. The constraint values (arch, mem, etc) are always the same. 
This pattern is repeated across many of our domains.